amqp/src/main/scala/akka/stream/alpakka/amqp/impl/AmqpPublishFlowStage.scala

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (C) 2016-2019 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.stream.alpakka.amqp.impl
+
+import akka.annotation.InternalApi
+import akka.stream.alpakka.amqp.{AmqpWriteSettings, WriteMessage}
+import akka.stream.stage.{GraphStage, GraphStageLogic, InHandler, OutHandler}
+import akka.stream._
+
+import scala.util.{Failure, Success, Try}
+
+/**
+ * Connects to an AMQP server upon materialization and sends incoming messages to the server.
+ * Each materialized flow will create one connection to the broker.
+ */
+@InternalApi
+private[amqp] final class AmqpPublishFlowStage[O](settings: AmqpWriteSettings)
+    extends GraphStage[FlowShape[(WriteMessage, O), O]] { stage =>
+
+  val in = Inlet[(WriteMessage, O)]("AmqpPublishConfirmFlow.in")
+  val out = Outlet[O]("AmqpPublishConfirmFlow.out")
+
+  override def shape: FlowShape[(WriteMessage, O), O] = FlowShape.of(in, out)
+
+  override protected def initialAttributes: Attributes =
+    Attributes
+      .name("AmqpPublishFlowStage")
+      .and(ActorAttributes.IODispatcher)
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new GraphStageLogic(shape) with AmqpConnectorLogic {
+      override val settings = stage.settings
+      private val exchange = settings.exchange.getOrElse("")
+      private val routingKey = settings.routingKey.getOrElse("")
+
+      override def whenConnected(): Unit =
+        if (settings.publishConfirm.isDefined) {
+          channel.confirmSelect()
+        }
+
+      setHandler(
+        out,
+        new OutHandler {
+          override def onPull(): Unit = pull(in)
+        }
+      )
+
+      setHandler(
+        in,
+        new InHandler {
+          override def onPush(): Unit = {
+            val (elem, passthrough) = grab(in)
+            channel.basicPublish(
+              exchange,
+              elem.routingKey.getOrElse(routingKey),
+              elem.mandatory,
+              elem.immediate,
+              elem.properties.orNull,
+              elem.bytes.toArray
+            )
+
+            settings.publishConfirm match {
+              case Some(publishConfirmConf) =>
+                Try(channel.waitForConfirmsOrDie(publishConfirmConf.confirmTimeout.toMillis)) match {
```

Comment:
I think that `channel.addConfirmListener` could be better - without blocking. What do you think?
You can copy it from #1584

Reply:
Well, I also thought of it at first, and it clearly as the advantage of being non-blocking and therefore faster, but I think the huge disadvantage of it is that it doesn't maintain order :/
With `waitForConfirmsOrDie`, it ensure order is maintained both in the flow and in amqp.

We'll see with usage, but I think use cases where `waitForConfirmsOrDie` is to slow are rather rare :/

What do you think ? Do you have such use cases ?

Reply:
Well, right now I have a code base in proprietary project (which is not written akka/alpakka) where exactly confirm listener is used. I wanted to migrate it to alpakka, but don't want to change the logic, because don't want to break things.

Order is not always needed. Probably can be restored by **deliveryTag** parameter, but not sure here.

Or may be this thing could be configurable and we can add both implementations. (can create next PR with adding new implementation and keeping this once this one will be merged).

Reply:
If I can add one thing here - waiting for confirmation after each publish (as with `waitForConfirmsOrDie`) is a true performance killer, especially in combination with persistent messages (more on that [here](https://www.rabbitmq.com/confirms.html#publisher-confirms-latency)). As @fancywriter mentioned, with `channel.addConfirmListener` order can be restored (next deliveryTag can be retrieved before message is published - I even started my own implementation this way), but actually I think both approaches are valuable and have their own use cases. It would be perfect to have 3 different flow connectors: 

- synchronous (like with `waitForConfirmsOrDie`) - preserving order of messages **delivered** to the message broker
- asynchronous ordered (with `channel.addConfirmListener` and some kind of output buffer that dispatches messages downstream in order) - more performant while preserving order in messages that **leave the graph stage**
- asynchronous unordered (also with `channel.addConfirmListener`) - the most performant but without any ordering guarantees

In absence of namming pattern in Alpakka for similar situation (at least I don't see one), those could be named similarly as variants of `map` operator from vanilla Akka Streams, i.e. `AmqpFlow.simple`, `AmqpFlow.async` and `AmqpFlow.asyncUnordered`

I also agree that we could just finish synchronous variant here, and remaining ones in another PR (I will be happy to contribute as well) - just please keep in mind other variants while naming stuff in public api ;)

Reply:
@kstrek Sounds like a plan. I hope the variants can be implemented by variants of this graph stage logic.

Reply:
Sorry for the late reply. Yes that seems like a great idea! 
Did you see anything in the current PR that would hinder future addition? It looked fine for me, but I may have missed some.

Reply:
@jcare44 I don't see anything in particular. Looks fine for me too.